src/Microsoft.Restier.Core/ApiBase.cs

```diff
@@ -85,7 +85,10 @@ protected ApiConfiguration Configuration
                     apiType =>
                     {
                         IServiceCollection services = new ServiceCollection()
-                            .CutoffPrevious<IQueryExecutor>(DefaultQueryExecutor.Instance);
+                            .CutoffPrevious<IQueryExecutor>(DefaultQueryExecutor.Instance)
+                            .AddSingleton<ApiConfigurationExtensions.PropertyBag>()
+                            .AddScoped<ApiContextExtensions.PropertyBag>()
+                            .AddScoped<InvocationContextExtensions.PropertyBag>();
```

Comment:
Perhaps I didn't get you exactly. IMHO we just declare a internal PropertyBag class and add it as scoped service, then ApiConfig, ApiContext, InvocationContext just get it from their own service provider.